packages/core/src/State.ts

```diff
@@ -298,7 +323,10 @@ export class State<
    * @param parentStateValue
    */
   public matches<TSV extends TTypestate['value']>(
-    parentStateValue: TSV
+    parentStateValue: TResolvedTypesMeta extends TypegenEnabled
+      ? // @ts-ignore
+        TResolvedTypesMeta['matchesStates']
```

Comment:
What's the problem here? (And in the other one)

Reply:
I don't know specifically with this one - might need a comment here.

Reply:
I will write more about the design of the whole thing (would love to describe this in a blog post as well) but to answer this question immediately:

The idea here is that I don't check if the `TResolvedTypesMeta extends TypegenMeta` because I trust the injected type (it is only supposed to be injected using our tooling). Checking all of the properties would impose some extra performance overhead - after all, you need to check if every property in the constraint type is satisfied by the provided/instantiated type. But since this never provided by the user we can only check a single property to see if we actually got a type with `'@@xstate/typegen': true` and continue assuming that the rest of this type is "OK".

I've used `// @ts-ignore` here since it felt cumbersome to use things like `Prop` and `Cast` all over the place to satisfy the compiler. It also felt that this might come with some perf boost since we wouldn't have to go through additional layers of types, check more constraints etc. I've tested this on a simple example, and `// @ts-ignore` performed better when it comes to the number of type instantiations than `Prop`. I've consulted this with Devan and he said that my example could be too simple and that `Prop` should perform better. It feels that this might be true but it also might depend on how sophisticated the actual implementation of TS is.

The `// @ts-ignore` approach has some drawbacks associated with it though. The problem is that TS can remove/misplace those comments when generating `.d.ts` and that in turn can raise errors in users' projects if they are not using `skipLibCheck: true`. They really should use that option but that's not our decision to make and we shouldn't enforce stuff like this on them - especially given the fact that the issue might be hard to diagnose on their end.

So I've decided that I will be changing all of those to `Prop`/`Cast` appropriately. The plan is to revisit the `// @ts-ignore` approach in the future though. I would love to test both versions on a real codebase - like XState Visualizer or smth - and make a more informed decision about it then. If the `// @ts-ignore` approach would come out of this experiment victorious then we could write a script that would include this comment at correct positions in the `.d.ts`.